sql/codegen_xgboost.go

```diff
@@ -0,0 +1,606 @@
+}
+
+func xgUnsupportedColTagError() error {
+	return fmt.Errorf("xgUnsupportedColTagError: valid column tags of xgboost engine([feature_columns, group, weight])")
+}
+
+func uIntPartial(key string, ptrFn func(*xgboostFiller) *uint) func(*map[string][]string, *xgboostFiller) error {
+	return func(a *map[string][]string, r *xgboostFiller) error {
+		// xgParseAttr will ensure the key is existing in map
+		val, _ := (*a)[key]
+		if len(val) != 1 {
+			return xgInvalidAttrValError(key, val)
+		}
+		if intVal, err := strconv.ParseUint(val[0], 10, 32); err != nil {
+			return xgInvalidAttrValError(key, val)
+		} else if intPtr := ptrFn(r); *intPtr != 0 {
+			return xgDupAttrSettingError(key)
+		} else {
+			*intPtr = uint(intVal)
+			delete(*a, key)
+		}
+		return nil
+	}
+}
+
+func fp32Partial(key string, ptrFn func(*xgboostFiller) *float32) func(*map[string][]string, *xgboostFiller) error {
+	return func(a *map[string][]string, r *xgboostFiller) error {
+		// xgParseAttr will ensure the key is existing in map
+		val, _ := (*a)[key]
+		if len(val) != 1 {
+			return xgInvalidAttrValError(key, val)
+		}
+		if fpVal, err := strconv.ParseFloat(val[0], 32); err != nil {
+			return xgInvalidAttrValError(key, val)
+		} else if fpPtr := ptrFn(r); *fpPtr != 0 {
+			return xgDupAttrSettingError(key)
+		} else {
+			*fpPtr = float32(fpVal)
+			delete(*a, key)
+		}
+		return nil
+	}
+}
+
+func boolPartial(key string, ptrFn func(*xgboostFiller) *bool) func(*map[string][]string, *xgboostFiller) error {
+	return func(a *map[string][]string, r *xgboostFiller) error {
+		// xgParseAttr will ensure the key is existing in map
+		val, _ := (*a)[key]
+		if len(val) != 1 {
+			return xgInvalidAttrValError(key, val)
+		}
+		bVal, err := strconv.ParseBool(val[0])
+		if err != nil {
+			return xgInvalidAttrValError(key, val)
+		}
+		bPtr := ptrFn(r)
+		*bPtr = bVal
+		delete(*a, key)
+		return nil
+	}
+}
+
+func strPartial(key string, ptrFn func(*xgboostFiller) *string) func(*map[string][]string, *xgboostFiller) error {
+	return func(a *map[string][]string, r *xgboostFiller) error {
+		// xgParseAttr will ensure the key is existing in map
+		val, _ := (*a)[key]
+		if len(val) != 1 {
+			return xgInvalidAttrValError(key, val)
+		}
+		stringPtr := ptrFn(r)
+		if len(*stringPtr) != 0 {
+			return xgDupAttrSettingError(key)
+		}
+		*stringPtr = val[0]
+		delete(*a, key)
+		return nil
+	}
+}
+
+func sListPartial(key string, ptrFn func(*xgboostFiller) *[]string) func(*map[string][]string, *xgboostFiller) error {
+	return func(a *map[string][]string, r *xgboostFiller) error {
+		// xgParseAttr will ensure the key is existing in map
+		val, _ := (*a)[key]
+		strListPtr := ptrFn(r)
+		if len(*strListPtr) != 0 {
+			return xgDupAttrSettingError(key)
+		}
+		*strListPtr = val
+		delete(*a, key)
+		return nil
+	}
+}
+
+var xgbAttrSetterMap = map[string]func(*map[string][]string, *xgboostFiller) error{
```

Comment:
Seems a simple way is to determine the attribute value type and set to an arbitrary map, then render the generated code. It's fine if the user setup the wrong attribute value type since it will throw errors at runtime.

Reply:
Yes, but it seems no harm to do some check during codegen?

Reply:
True. I personally like simpler codes, it's your choice, this do not block merging.

Reply:
True. I personally like simpler codes, it's your choice, this do not block merging.